plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx

```diff
@@ -0,0 +1,288 @@
+/**
+ * External dependencies
+ */
+import React, { useRef, useState } from 'react';
+import classNames from 'classnames';
+import {
+	useSelect,
+	UseSelectState,
+	UseSelectStateChangeOptions,
+} from 'downshift';
+import { Button } from '@wordpress/components';
+import { useCallback } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+import { check, chevronDown, Icon } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import { Item, ControlProps } from './types';
+import { getOptionLabel } from './utils';
+import './country-selector.scss';
+
+// State reducer to control selection navigation
+const stateReducer = < ItemType extends Item >(
+	state: UseSelectState< ItemType | null >,
+	actionAndChanges: UseSelectStateChangeOptions< ItemType | null >
+): Partial< UseSelectState< ItemType > > => {
+	const { changes, type } = actionAndChanges;
+	switch ( type ) {
+		case useSelect.stateChangeTypes.ItemClick:
+			return {
+				...changes,
+				isOpen: true, // Keep menu open after selection.
+				highlightedIndex: state.highlightedIndex,
+			};
+		default:
+			return changes;
+	}
+};
+
+export const CountrySelector = < ItemType extends Item >( {
+	name,
+	className,
+	label,
+	describedBy,
+	options: items,
+	onChange,
+	value,
+	placeholder,
+	children,
+}: ControlProps< ItemType > ): JSX.Element => {
+	const [ searchText, setSearchText ] = useState( '' );
+	const [ isSearchFocused, setSearchFocused ] = useState( false );
+	const [ visibleItems, setVisibleItems ] = useState(
+		new Set( [ ...items.map( ( i ) => i.key ) ] )
+	);
+
+	const itemsToRender = items.filter( ( item ) =>
+		visibleItems.has( item.key )
+	);
+
+	const {
+		getToggleButtonProps,
+		getMenuProps,
+		getItemProps,
+		isOpen,
+		highlightedIndex,
+		selectedItem,
+		closeMenu,
+		selectItem,
+	} = useSelect< ItemType >( {
+		initialSelectedItem: value,
+		items: itemsToRender,
+		stateReducer,
+		onIsOpenChange: () => selectItem( value ),
+	} );
+
+	const itemString = getOptionLabel( value.key, items );
+	const selectedValue = selectedItem ? selectedItem.key : '';
+
+	const searchRef = useRef< HTMLInputElement >( null );
+	const previousStateRef = useRef< {
+		visibleItems: Set< string >;
+	} >();
+
+	function getDescribedBy() {
+		if ( describedBy ) {
+			return describedBy;
+		}
+
+		if ( ! itemString ) {
+			return __( 'No selection', 'woocommerce' );
+		}
+
+		return sprintf(
+			// translators: %s: The selected option.
+			__( 'Currently selected: %s', 'woocommerce' ),
+			itemString
+		);
+	}
+
+	const getSearchSuffix = ( focused: boolean ) => {
+		if ( focused ) {
+			return (
+				<svg
```

Comment:
Would you consider moving the svg out either into a .svg asset or into another named react component if that's not a good option? Having this svg inline makes it hard to read and also doesn't give an idea of what image it represents

Reply:
Done!